regional_mom6/rotation.py

```diff
@@ -0,0 +1,342 @@
+        q_points.isel(nyp=slice(0, -1), nxp=slice(1, None)),
+        t_points,
+    )
+
+
+def modulo_around_point(x, x0, L):
+    """
+    Return the modulo-:math:`L` value of :math:`x` within the interval :math:`[x_0 - L/2, x_0 + L/2]`.
+    If :math:`L ≤ 0`, then we get back :math:`x`.
+
+    (Adapted from MOM6 code; https://github.com/mom-ocean/MOM6/blob/776be843e904d85c7035ffa00233b962a03bfbb4/src/initialization/MOM_shared_initialization.F90#L592-L606)
+
+    Parameters
+    ----------
+    x: float
+       Value to which to apply modulo arithmetic
+    x0: float
+        Center of modulo range
+    L: float
+       Modulo range width
+
+    Returns
+    -------
+    float
+        ``x`` shifted by an integer multiple of ``L`` to be closer to ``x0``, i.e., within the interval ``[x0 - L/2, x0 + L/2]``
+    """
+    if L <= 0:
+        return x
+    else:
+        return ((x - (x0 - L / 2)) % L) + (x0 - L / 2)
+
+
+def mom6_angle_calculation_method(
+    len_lon,
+    top_left: xr.DataArray,
+    top_right: xr.DataArray,
+    bottom_left: xr.DataArray,
+    bottom_right: xr.DataArray,
+    point: xr.DataArray,
+) -> xr.DataArray:
+    """
+    Calculate the angle of the point using the MOM6 method in :func:`~initialize_grid_rotation_angle`.
+    This method can handle vectorized computations.
+
+    (Adapted from MOM6 code; https://github.com/mom-ocean/MOM6/blob/e818ea4e792f0b85797247f955789b3c1210db8d/src/initialization/MOM_shared_initialization.F90#L535)
+
+    Parameters
+    ----------
+    len_lon: float
+        The length of the longitude of the regional domain
+    top_left, top_right, bottom_left, bottom_right: xr.DataArray
+        The four points around the point to calculate the angle from the hgrid requires an x and y component
+    point: xr.DataArray
+        The point to calculate the angle from the hgrid
+
+    Returns
+    -------
+    xr.DataArray
+        The angle of the point
+    """
+    rotation_logger.info("Calculating grid rotation angle")
+    # Direct Translation
+    pi_720deg = (
+        np.arctan(1) / 180
+    )  # One quarter the conversion factor from degrees to radians
+
+    # Compute lonB for all points
+    lonB = np.zeros((2, 2, len(point.nyp), len(point.nxp)))
+
+    # Vectorized computation of lonB
+    lonB[0][0] = modulo_around_point(bottom_left.x, point.x, len_lon)  # Bottom Left
+    lonB[1][0] = modulo_around_point(top_left.x, point.x, len_lon)  # Top Left
+    lonB[1][1] = modulo_around_point(top_right.x, point.x, len_lon)  # Top Right
+    lonB[0][1] = modulo_around_point(bottom_right.x, point.x, len_lon)  # Bottom Right
+
+    # Compute lon_scale
+    lon_scale = np.cos(
+        pi_720deg * ((bottom_left.y + bottom_right.y) + (top_right.y + top_left.y))
+    )
+
+    # Compute angle
+    angle = np.arctan2(
+        lon_scale * ((lonB[0, 1] - lonB[1, 0]) + (lonB[1, 1] - lonB[0, 0])),
+        (bottom_left.y - top_right.y) + (top_left.y - bottom_right.y),
```

Comment:
if we look at

https://github.com/mom-ocean/MOM6/blob/05d8cc395c1c3c04dd04885bf8dd6df50a86b862/src/initialization/MOM_shared_initialization.F90#L581

it should be

```suggestion
        (top_right.y - bottom_left.y) + (top_left.y - bottom_right.y),
```

right @manishvenu, @ashjbarnes?

Reply:
Yes it is, good catch, thanks!

Reply:
I can do a little bit of testing next week for this

Reply:
Getting weird results, let me look at this further

Reply:
Ok I figured it out, this bug worked the first time because the numerator is also flipped the exact same way lonB is in indexing [y,x] so the total line should be : 

`    angle = np.arctan2(
        lon_scale * (( lonB[1, 0] - lonB[0, 1]) + (lonB[1, 1] - lonB[0, 0])),
        (top_right.y - bottom_left.y) + (top_left.y - bottom_right.y)
    )`

Basically, you always want the top points subtracting the bottom points on a diagonal.

We either do the suggestion here, or change the lonB matrix to be x,y instead of y, x. I can make this change.

Interesting note, because they were both flipped, thankfully, there were only differences in angles up to the 1e-5

Reply:
how's the commit I pushed?

the difference in angles "only up to 1e-5" depends on the grid you are testing; coarser grid might have more difference!

Reply:
Great - Fixed it. Forgot to swap the other part of the numerator

Reply:
I had a 90 degree transformation, didn't need that

Reply:
This seems resolved to me!